content/articles/mocking-with-junit-and-mockito-the-why-and-how/index.md

```diff
@@ -0,0 +1,871 @@
+        for (StudentDto dto: registeredStudents) {
+            studentService.assignBedSpace(dto);
+        }
+        List<String> studentsInHall3 = studentService.returnNamesOfAllStudentsInAHostel("HALL3");
+        List<String> studentsInHall1 = studentService.returnNamesOfAllStudentsInAHostel("HALL1");
+
+        assertThat(studentsInHall3, hasSize(2));
+        assertThat(studentsInHall3, hasItems("John Doe", "Peter Rabbit"));
+        assertThat(studentsInHall1, hasSize(2));
+        assertThat(studentsInHall1, hasItems("Mary Fallow", "Patricia Lemon"));
+    }
+
+  private List<StudentDto> registerStudents() throws Exception {
+    RegistrationRequest firstRegistrationRequest = new RegistrationRequest("John", "Doe", "securedPassword", "MAT100419", Gender.MALE);
+    RegistrationRequest secondRegistrationRequest = new RegistrationRequest("Mary", "Fallow", "securedWassword", "MAT100420", Gender.FEMALE);
+    RegistrationRequest thirdRegistrationRequest = new RegistrationRequest("Peter", "Rabbit", "securedDassword", "MAT100421", Gender.MALE);
+    RegistrationRequest fourthRegistrationRequest = new RegistrationRequest("Patricia", "Lemon", "securedBassword", "MAT100422", Gender.FEMALE);
+    StudentDto firstStudentDto = studentService.registerStudent(firstRegistrationRequest);
+    StudentDto secondStudentDto = studentService.registerStudent(secondRegistrationRequest);
+    StudentDto thirdStudentDto = studentService.registerStudent(thirdRegistrationRequest);
+    StudentDto fourthStudentDto = studentService.registerStudent(fourthRegistrationRequest);
+
+    return new ArrayList<>(List.of(firstStudentDto, secondStudentDto,
+      thirdStudentDto, fourthStudentDto));
+  }
+}
+```
+
+When we *spy* on a class, Mockito spins an actual class object but now interactions with the object can be logged.
+
+> Notice that we have removed the ```when -then``` directives from our test case.
+We refactor the test case for the ```registerStudent``` method as follows:
+
+```java
+ @Test
+    void registerStudentTest() throws Exception {
+      RegistrationRequest registrationRequest = new RegistrationRequest("John", "Fallow",
+      "my_password", "PSC1004396", Gender.MALE);
+      Student student = Student.builder()
+                    .firstName("John")
+                    .lastName("Fallow")
+                    .matricNo("PSC100419")
+                    .password("my_password")
+                    .gender(Gender.MALE).build();
+      doReturn(student).when(studentRepository).save(any(Student.class));
+      StudentDto studentDto = studentService.registerStudent(registrationRequest);
+      assertThat(studentDto, hasProperty("firstName", equalTo("John")));
+      assertThat(studentDto, hasProperty("lastName", equalTo("Fallow")));
+      assertThat(studentDto, hasProperty("matricNo", equalTo("PSC100419")));
+      assertThat(studentDto, hasProperty("gender", equalTo(Gender.MALE)));
+  }
+```
```

Comment:
Also this one, the identation seems to have moved abit, make it this way:

```java
@Test
void registerStudentTest() throws Exception {
    RegistrationRequest registrationRequest = new RegistrationRequest("John", "Fallow","my_password", "PSC1004396", Gender.MALE);
    Student student = Student.builder()
                        .firstName("John")
                        .lastName("Fallow")
                        .matricNo("PSC100419")
                        .password("my_password")
                        .gender(Gender.MALE).build();
    doReturn(student).when(studentRepository).save(any(Student.class));
    StudentDto studentDto = studentService.registerStudent(registrationRequest);
    assertThat(studentDto, hasProperty("firstName", equalTo("John")));
    assertThat(studentDto, hasProperty("lastName", equalTo("Fallow")));
    assertThat(studentDto, hasProperty("matricNo", equalTo("PSC100419")));
    assertThat(studentDto, hasProperty("gender", equalTo(Gender.MALE)));
}
```

Reply:
Done